app/src/middleware/apiTracker.js

```diff
@@ -0,0 +1,185 @@
+        try {
+            if (restArgs[0]) {
+                chunks.push(Buffer.from(restArgs[0]));
+            }
+            const body = Buffer.concat(chunks).toString('utf8');
+            const obj = JSON.parse(body);
+
+        } catch (err) {
+            log.error('mailApiTracker', err);
+        }
+        defaultEnd.apply(res, restArgs);
+    };
+    */
+  }
+  next();
+};
+
+const initializeApiTracker = (app) => {
+
+  // register token parser functions.
+  // this one would depend on authorizedParty middleware being loaded
+  morgan.token('azp', (req) => {
+    return req.authorizedParty ? req.authorizedParty : '-';
+  });
+
+  morgan.token('operation', (req) => {
+    return req._operation ? req._operation : '-';
+  });
+
+  morgan.token('timestamp', (req) => {
+    return req._timestamp ? req._timestamp : '-';
+  });
+
+  morgan.token('contextKeyCount', (req) => {
+    function countKeys(source) {
+      if (!source) return 0;
+      let result = 0;
+      (function count(obj) {
+        if (Array.isArray(obj)) {
+          obj.forEach(function (j) {
+            count(j);
+          });
+        } else {
+          Object.keys(obj).forEach(function (k) {
+            result++;
+            const v = obj[k];
+            if (typeof v === 'object') {
+              count(v);
+            }
+          });
+        }
+      })(source);
+      return result;
+    }
+
+    try {
+      // want to return some idea of the size/magnitude/complexity of the contexts
+      return countKeys(req.body.contexts);
+    } catch (e) {
+      return '0';
+    }
+  });
+
+  morgan.token('contentFileType', (req) => {
+    try {
+      return req.body.template.contentFileType;
+    } catch (e) {
+      return '-';
+    }
+  });
+
+  morgan.token('contentEncodingType', (req) => {
+    try {
+      return req.body.template.contentEncodingType;
+    } catch (e) {
+      return '-';
+    }
+  });
+
+  morgan.token('contentSize', (req) => {
+    try {
+      const buf = Buffer.from(req.body.template.content, req.body.template.contentEncodingType);
+      return buf.length;
+    } catch (e) {
+      return '0';
+    }
+  });
+
+  morgan.token('outputFileType', (req) => {
+    try {
+      return req.body.template.outputFileType;
+    } catch (e) {
+      return '-';
+    }
+  });
+
+  app.use(morgan(apiTrackerFormat, {
+    // eslint-disable-next-line no-unused-vars
+    skip: function (req, res) {
+      return !trackerUrls.includes(req.baseUrl);
+    },
+    stream: {
+      write: (s) => {
+        if (s && s.trim().length > 0) {
+          const parts = s.trim().split(' ');
+          const ts = Number.parseInt(parts[2]);
+          // format this to match Kibana pretty format (almost - utc + timezone)...
+          const timestamp = moment.utc(ts).format('MMMM Do YYYY, HH:mm:ss.SSSZZ');
+          const o = {
+            clogs: {
+              type: 'CDOGS_API_TRACKER',
+              ts: ts,
+              timestamp: timestamp,
```

Comment:
Is there a reason why we should have 2 timestamps/ts? I can see it'd be useful as an alias for being able to more easily select things, but I can also see some potential confusion if they don't end up meaning/representing the same thing. It might be good to just pick one of the two (maybe just tz) and run with that for now?

Reply:
lucas wanted a readable string, not just unix number